inference-engine/thirdparty/clDNN/src/gpu/device_info.cpp

```diff
@@ -26,11 +26,121 @@
+#include <dlfcn.h>
+#endif
 
 namespace cldnn {
 namespace gpu {
 
 namespace {
+static constexpr auto device_info_failed_msg = "Device lookup failed";
+
+int driver_dev_id()
+{
+    const std::vector<int> unused_ids = {
+        0x4905, 0x4906, 0x4907, 0x4908
+    };
+    std::vector<int> result;
+
+#ifdef _WIN32
+    {
+        HDEVINFO device_info_set = SetupDiGetClassDevsA(&GUID_DEVCLASS_DISPLAY, NULL, NULL, DIGCF_PRESENT);
+        if (device_info_set == INVALID_HANDLE_VALUE)
+            return 0;
+
+        SP_DEVINFO_DATA devinfo_data;
+        std::memset(&devinfo_data, 0, sizeof(devinfo_data));
+        devinfo_data.cbSize = sizeof(devinfo_data);
+
+        for (DWORD dev_idx = 0; SetupDiEnumDeviceInfo(device_info_set, dev_idx, &devinfo_data); dev_idx++) {
+            const size_t buf_size = 512;
+            char buf[buf_size];
+            if (!SetupDiGetDeviceInstanceIdA(device_info_set, &devinfo_data, buf, buf_size, NULL)) {
+                continue;
+            }
+
+            char* vendor_pos = std::strstr(buf, "VEN_");
+            if (vendor_pos != NULL && std::stoi(vendor_pos + 4, NULL, 16) == 0x8086) {
+                char* device_pos = strstr(vendor_pos, "DEV_");
+                if (device_pos != NULL) {
+                    result.push_back(std::stoi(device_pos + 4, NULL, 16));
+                }
+            }
+        }
+
+        if (device_info_set) {
+            SetupDiDestroyDeviceInfoList(device_info_set);
+        }
+    }
+#elif defined(__linux__)
+    {
+        std::string dev_base{ "/sys/devices/pci0000:00/0000:00:02.0/" };
+        std::ifstream ifs(dev_base + "vendor");
+        if (ifs.good())
+        {
+            int ven_id;
+            ifs >> std::hex >> ven_id;
+            ifs.close();
+            if (ven_id == 0x8086)
+            {
+                ifs.open(dev_base + "device");
+                if (ifs.good())
+                {
+                    int res = 0;
+                    ifs >> std::hex >> res;
+                    result.push_back(res);
+                }
+            }
+        }
+    }
+#endif
+
+    auto id_itr = result.begin();
+    do {
+        if (std::find(unused_ids.begin(), unused_ids.end(), *id_itr) != unused_ids.end())
+            result.erase(id_itr);
+        else
+            id_itr++;
+    } while (id_itr != result.end());
+
+    return result.back();
+}
+
+bool get_imad_support(const cl::Device& device) {
+    std::string dev_name = device.getInfo<CL_DEVICE_NAME>();
+
+    if (dev_name.find("Gen12") != std::string::npos)
+        return true;
+
+    auto flag = device.getInfo<CL_DEVICE_HOST_UNIFIED_MEMORY>();
+    if (flag != 0) {
+        const std::vector<int> imad_ids = {
+            0x9A40, 0x9A49, 0x9A59, 0x9AD9,
+            0x9A60, 0x9A68, 0x9A70, 0x9A78,
+            0x9A7F, 0x9AF8, 0x9AC0, 0x9AC9
+        };
+        int dev_id = driver_dev_id();
+        if (dev_id == 0) {
+            throw std::runtime_error(device_info_failed_msg);
```

Comment:
Maybe throwing an exception is a bit too much in this case? Can we just return false and use default path?

Reply:
Fixed